core/src/main/java/org/apache/calcite/jdbc/SimpleCalciteSchema.java

```diff
@@ -70,28 +69,29 @@ public CalciteSchema add(String name, Schema schema) {
   protected CalciteSchema getImplicitSubSchema(String schemaName,
       boolean caseSensitive) {
     // Check implicit schemas.
-    Schema s = schema.getSubSchema(schemaName);
-    if (s != null) {
-      return new SimpleCalciteSchema(this, s, schemaName);
+    final NameSet names = NameSet.immutableCopyOf(schema.getSubSchemaNames());
+    for (String schemaName2: names.range(schemaName, caseSensitive)) {
+      return new SimpleCalciteSchema(this,
+          schema.getSubSchema(schemaName2), schemaName);
     }
     return null;
   }
 
   protected TableEntry getImplicitTable(String tableName,
       boolean caseSensitive) {
     // Check implicit tables.
-    Table table = schema.getTable(tableName);
-    if (table != null) {
-      return tableEntry(tableName, table);
+    final NameSet names = NameSet.immutableCopyOf(schema.getTableNames());
+    for (String tableName2: names.range(tableName, caseSensitive)) {
```

Comment:
Is this acceptable for performance?
Copying all the tables does not look OK for each `get...` call.

Reply:
Thanks, you are right, the performance is not acceptable.
We should not copy them and use `NameSet` as that in `CachingCalciteSchema`.
I implement a `lookup` method and make some performance tests in my local environment.
```
tables            NameSet    lookup
100000         224ms        9ms
1000000       1639ms      21ms
10000000     14069ms    117ms
```

Reply:
I would go one step further and not even call `getTableNames()` etc., if it is not necessary. In some cases `schema.getTableNames()` may be constructed each time on demand.

Reply:
Could you elaborate it? Or do you mean enumerating all possibilities?
E.g., converting `getTable("xy", false) `  to
`getTable("XY") + getTable("Xy") + getTable("xY") +getTable("xy")`?

Reply:
Thanks! I see, ignore the above comment.